feat: Allow specifying the user agent per-sdk implementation.

kinyoklion opened the pull request:
Currently `NodeJSClient` is being used as the user agent for all server SDKs.

This was not apparent because only node was sending events. Events are being added to Vercel and it revealed this issue.

kinyoklion commented on packages/shared/akamai-edgeworker-sdk/src/platform/info/index.ts:
```diff
@@ -17,6 +17,7 @@ class AkamaiPlatformInfo implements Info {
     return {
       name: this.sdkName,
       version: this.sdkVersion,
+      userAgent: 'Akamai',
```
@ctawiah or @dariushar What should the user agent be for Akamai? It isn't sending events, but if it ever does, then it may matter.

ctawiah replied:
Our integration userAgent typically follow this pattern `launchdarkly/{integration-name}` so i suggested an update to the name.

kinyoklion replied:
Ok, so generally speaking these things we know are launchdarkly. So we have not had launchdarkly in the names before. They have not been terribly consistent. 

Like node is "NodeJSClient". We typically have had client/server in them.

I am leaning toward making it the same as the vercel one. AkamaiEdgeSDK.

ctawiah replied:
Ok, that works.

kinyoklion commented on packages/sdk/vercel/src/createPlatformInfo.ts:
```diff
@@ -11,6 +11,7 @@ class VercelPlatformInfo implements Info {
     return {
       name: '@launchdarkly/vercel-server-sdk',
       version: '__LD_VERSION__',
+      userAgent: 'VercelEdgeSDK',
```
@ldhenry Once this is merged, then you won't need the wrapper name/version.

kinyoklion commented on packages/shared/sdk-server/src/data_sources/defaultHeaders.ts:
```diff
@@ -12,7 +12,7 @@ export default function defaultHeaders(
   const headers: { [key: string]: string } = {
     authorization: sdkKey,
-    'user-agent': `NodeJSClient/${sdkData.version}`,
+    'user-agent': `${sdkData.userAgent ? sdkData.userAgent : 'NodeJSClient'}/${sdkData.version}`,
```
Just uses the default for compatibility. Only really matters if someone uses a newer version of this package than their SDK shipped with.

louis-launchdarkly reviewed (commented):
I see a test failure about Akamai, and there already has a comment discussing Akamai. The other part looks good

kinyoklion commented:
> I see a test failure about Akamai, and there already has a comment discussing Akamai. The other part looks good

Should be resolved now. I'll wait to merge/release until I hear from some of the edge SDK folks.

ctawiah commented on packages/shared/akamai-edgeworker-sdk/src/platform/info/index.ts:
```diff
@@ -17,6 +17,7 @@ class AkamaiPlatformInfo implements Info {
     return {
       name: this.sdkName,
       version: this.sdkVersion,
+      userAgent: 'Akamai',
```
```suggestion
      userAgent: 'launchdarkly/akamai',
```

kinyoklion replied:
So. the user agent gets formed with a version. So this would be launchdarkly/akamai/version. I think this would cause problems with our event pipeline.

ctawiah replied:
oh hmm, in the case we do `launchdarkly-akamai`

yusinto commented on packages/sdk/server-node/src/platform/NodeInfo.ts:
```diff
@@ -37,6 +37,7 @@ export default class NodeInfo implements platform.Info {
     return {
       name: packageJson.name,
       version: packageJson.version,
+      userAgentBase: 'NodeJSClient',
```
I am curious if we can just use `packageJson.name` instead? If we do then the useragent string will look like `@launchdarkly/js-server-sdk-common/1.0.6` which is longer, but it's ok?

kinyoklion replied:
Well, this one needs to specifically be `NodeJSClient`. Because it not a new user agent name and there isn't a good reason to create new statistics.

The second thing, as I mentioned in another comment is, what would be the effect of this be on our pipeline. All of our user agents are of the form `thing/version`.

yusinto commented on packages/shared/sdk-server/src/data_sources/defaultHeaders.ts:
```diff
@@ -12,7 +12,9 @@ export default function defaultHeaders(
   const headers: { [key: string]: string } = {
     authorization: sdkKey,
-    'user-agent': `NodeJSClient/${sdkData.version}`,
+    'user-agent': `${sdkData.userAgentBase ? sdkData.userAgentBase : 'NodeJSClient'}/${
+      sdkData.version
+    }`,
```
Eslint complained about using unnecessary ternary here, so to fix that:

```suggestion
'user-agent': `${userAgentBase ?? 'NodeJSClient'}/${version}`,
```

husseinsakr commented:
Does this mean we can now set the userAgent from the node server sdk? or is this internal stuff?

kinyoklion commented:
> Does this mean we can now set the userAgent from the node server sdk? or is this internal stuff?

It is internal stuff basically.

Our edge SDKs used to be wrappers around our node SDK, but when we moved to this repo we refactored that. We wanted to remove any node dependencies from the edge SDKs to prevent the need for a large number of polyfills. So we moved to a model with a common library with platform implementations per SDK.

This change allowed those SDKs to change the user agent.

What is the use case you have for changing the user agent? 

Thank you,
Ryan

husseinsakr commented:
Hi @kinyoklion , thank you for the reply. I missed your comment. 

Our use case is that it we today deploy the launchdarkly relay proxy in our environments and have wrapped that solution as well. Our clients connect to it but we have no visibility from the relay proxy on who is connected to it. 

I figured if we are able to set a custom userAgent in our clients. We can tell from the relay proxy which clients are connected to it by reading from the header.

kinyoklion commented:
@husseinsakr Yeah, this wouldn't really help your use case.

The user-agent, is used for some features, like reporting out of date SDKs or available upgrades, so changing it wouldn't be great.

In some SDKs we do allow custom headers to be added, which would potentially solve your problem. If you added your own custom header, then you could use it to collect such information.

This SDK doesn't support it, but it would likely be the right feature to request. Or there could be some more relay specific option that could help. You may wish to discuss it with support.

Thanks,
Ryan
